plugins/wacom-usb/fu-wac-module-bluetooth-id9.c

```diff
@@ -0,0 +1,261 @@
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CMD_NORMAL	   0x00
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CMD_FULLERASE  0xFE
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_RAM	   0x02
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_BEGIN   0x03
+#define FU_WAC_MODULE_BLUETOOTH_ID9_LOADER_DATA	   0x04
+#define FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL 0xEDB88320
+
+typedef struct __attribute__((packed)) {
+	guint16 unknown1;
+	guint32 unknown2;
+	guint16 size;
+} FuWacModuleBluetoothId9UnknownCmd;
+
+typedef struct __attribute__((packed)) {
+	guint8 spi_cmd;
+	guint32 start_addr;
+	guint16 size;
+	FuWacModuleBluetoothId9UnknownCmd data;
+
+} FuWacModuleBluetoothId9SpiCmd;
+
+typedef struct __attribute__((packed)) {
+	guint8 command;
+	guint16 size;
+	guint32 crc;
+	FuWacModuleBluetoothId9SpiCmd data;
+} FuWacModuleBluetoothId9LoaderCmd;
+
+static GBytes *
+fu_wac_module_bluetooth_id9_get_startcmd(const guint8 *data, gsize sz, gboolean full_erase)
+{
+	FuWacModuleBluetoothId9LoaderCmd cmd = {
+	    .command = full_erase ? FU_WAC_MODULE_BLUETOOTH_ID9_CMD_FULLERASE
+				  : FU_WAC_MODULE_BLUETOOTH_ID9_CMD_NORMAL,
+	    .size = GUINT16_TO_BE((guint16)(sizeof(cmd.data) + sz)),
+	    .crc = 0,
+	    .data =
+		{
+		    .spi_cmd = 0x91,
+		    .start_addr = 0,
+		    .size = GUINT16_TO_BE((guint16)(sizeof(cmd.data.data) + sz)),
+		    .data =
+			{
+			    .unknown1 = GUINT16_TO_BE((guint16)0x7050),
+			    .unknown2 = 0,
+			    .size = GUINT16_TO_BE((guint16)sz),
+			},
+		},
+	};
+	guint32 crc;
+
+	// CRC(append(cmd.data, *data))
+	crc = fu_crc32_full((const guint8 *)&cmd.data,
+			    sizeof(cmd.data),
+			    ~0,
+			    FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL);
+	crc = fu_crc32_full(data, sz, ~crc, FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL);
+	cmd.crc = GUINT32_TO_BE(crc);
+
+	return g_bytes_new(&cmd, sizeof(cmd));
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_block(FuWacModule *self,
+					guint8 phase,
+					const guint8 *data,
+					gsize sz,
+					FuProgress *progress,
+					GError **error)
+{
+	g_autoptr(GBytes) blob_chunk = NULL;
+	gsize buf_sz = sz + 1;
+	g_autofree guint8 *buf = g_malloc(buf_sz);
+
+	if (buf == NULL)
+		return FALSE;
+
+	buf[0] = phase;
+	memcpy(buf + 1, data, sz);
+
+	blob_chunk = g_bytes_new_take(g_steal_pointer(&buf), buf_sz);
+	if (!fu_wac_module_set_feature(self,
+				       FU_WAC_MODULE_COMMAND_DATA,
+				       blob_chunk,
+				       fu_progress_get_child(progress),
+				       FU_WAC_MODULE_WRITE_TIMEOUT,
+				       error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_blocks(FuWacModule *self,
+					 guint8 phase,
+					 const guint8 *data,
+					 gsize sz,
+					 gsize block_len,
+					 FuProgress *progress,
+					 GError **error)
+{
+	for (gsize i = 0; i < sz; i += block_len) {
+		const gsize remain = sz - i;
+		const gsize copy_len = remain < block_len ? remain : block_len;
+
+		fu_progress_set_percentage_full(fu_progress_get_child(progress), i, sz);
```

Comment:
nah, can you use fu_progress_set_steps() and fu_progress_step_done() please -- then the daemon knows what's coming.